SmellDetector/Detectors/RuleEngines/ClassMetricRuleEngine.cs

```diff
@@ -10,34 +11,73 @@ namespace SmellDetector.Detectors.RuleEngines
     internal class ClassMetricRuleEngine : IDetector
     {
         private readonly List<Rule> _rules;
+        private readonly List<Rule> _dynamicRules;
         public ClassMetricRuleEngine()
         {
             Rule rule1 = new Rule("10.1109/WCRE.2005.15",
                                   new AndCriteria(
                                             new AndCriteria(
-                                                new MetricCriteria("ATFD", OperationEnum.GREATER_THAN, 2),
-                                                new MetricCriteria("WMC", OperationEnum.GREATER_THAN, 47)),
-                                            new MetricCriteria("TCC", OperationEnum.LESS_THAN, 0.33)),
+                                                new MetricCriteria(CaDETMetric.ATFD , OperationEnum.GREATER_THAN, 2),
+                                                new MetricCriteria(CaDETMetric.WMC, OperationEnum.GREATER_THAN, 47)),
+                                            new MetricCriteria(CaDETMetric.TCC, OperationEnum.LESS_THAN, 0.33)),
                                   SmellType.GOD_CLASS);
             Rule rule2 = new Rule("10.1109/SCAM.2013.6648192",
                                   new OrCriteria(
-                                      new MetricCriteria("LOC", OperationEnum.GREATER_THAN, 750),
-                                      new MetricCriteria("[NOM] + [NOF]", OperationEnum.GREATER_THAN, 20)
+                                      new MetricCriteria(CaDETMetric.CLOC, OperationEnum.GREATER_THAN, 750),
+                                      new MetricCriteria(CaDETMetric.NMD_NAD, OperationEnum.GREATER_THAN, 20)
                                       ),
                                   SmellType.GOD_CLASS);
             Rule rule3 = new Rule("10.1109/MSR.2007.21",
                                   new OrCriteria(
-                                            new MetricCriteria("NMD", OperationEnum.GREATER_THAN, 15),
-                                            new MetricCriteria("NAD", OperationEnum.GREATER_THAN, 15)),
+                                            new MetricCriteria(CaDETMetric.NMD, OperationEnum.GREATER_THAN, 15),
+                                            new MetricCriteria(CaDETMetric.NAD, OperationEnum.GREATER_THAN, 15)),
                                   SmellType.GOD_CLASS);
             _rules = new List<Rule>();
             _rules.Add(rule1);
             _rules.Add(rule2);
             _rules.Add(rule3);
+            _dynamicRules = new List<Rule>();
+        }
+
+        private void createDynamicRules(List<CaDETClassDTO> caDetClassDtoList)
+        {
+            int indexOfSignificativeMetricValue = calculateIndexBasedOnPercentage(caDetClassDtoList, 20);
+            double atfdThreshold = calculateDynamicThreshold(caDetClassDtoList, CaDETMetric.ATFD, indexOfSignificativeMetricValue);
+
+            Rule rule1 = new Rule("10.1016/j.jss.2006.10.018",
```

Comment:
We should definitely create a rule parser next, so that we can serialize and deserialize these rules and make them far more flexible.

Once we conclude this PR it should be our next priority, so that we can move these rules to a file (e.g., a row will look like "10.1016/j.jss.2006.10.018; (ATFD > 10% AND ATFD > 4) AND (WMC > 20 AND TCC < 0.33)", and do the same for the long method rules. We should think about the special cases "e.g., top 10%, top 10 instances etc." and how we want to represent them in storage.

This would require a new class, e.g., RuleTextSerializier.